### PR TITLE
Fix 1810 suggest artifact assembly

### PR DESCRIPTION
**Pull request by vmarkovtsev**
Try to fix http://bugs.vcmi.eu/view.php?id=1810

### CODE REVIEW

File: client/widgets/CArtifactHolder.cpp [hunk 1]
```diff
@@ -214,6 +214,33 @@ void CArtPlace::clickLeft(tribool down, bool previousState)
 	}
 }
 
+bool CArtPlace::askToAssemble(const CArtifactInstance *art, ArtifactPosition slot,
+                              const CGHeroInstance *hero)
+{
+  std::vector<const CArtifact *> assemblyPossibilities = art->assemblyPossibilities(hero);
```
Comment (alexvins) [inline client/widgets/CArtifactHolder.cpp]: Use TAB fro indentation.

File: client/widgets/CArtifactHolder.cpp [hunk 2]
```diff
@@ -214,6 +214,33 @@ void CArtPlace::clickLeft(tribool down, bool previousState)
...
+      true,
+      std::bind(&CCallback::assembleArtifacts, LOCPLINT->cb.get(), hero, slot, true, combination->id),
+      0);
+
+    if(assemblyPossibilities.size() > 2)
+    {
+      logGlobal->warnStream() << boost::format(
+        "More than one possibility of assembling on %s... taking only first")
+        % art->artType->Name();
+      break;
```
Comment (alexvins) [inline client/widgets/CArtifactHolder.cpp]: Function will return false if "More than one possibility", is it intended?

**Comment by ArseniyShestakov** [comment]
Other than spaces everywhere I only see one thing need to be changed. When you pick last part of combined artifact from map window with suggestion to combine them appear before the info window about picked artifact.

Other unrelated problem I noticed is that "CGameHandler::assembleArtifacts" don't return true so client think that combine artifact package wasn't applied. Though we may need to check all other server code for same issue.

**Comment by vmarkovtsev** [comment]
Yeah, I spent an hour trying to make the windows be sequentail. With current arch, this looks impossible. You see, when something is picked, info window is shown asynchronously, and artifactPut event is received independently on that window close. To fix this, one has to refactor much in CGArtifact.onHeroVisit, IGameCallback, etc. to create a blocking InfoWindow with close event handling. I suggest to create a separate issue for this, because graceless combo suggestion is better than no suggestion at all.

**Comment by ArseniyShestakov** [comment]
Yep I suspected that may be tricky and I do agree that not perfect feature is better than no feature in that case.

**Comment by vmarkovtsev** [comment]
Thanks for CGameHandler::assembleArtifacts! Changed it to return true. I wondered why that "fishy" messages appear.

File: client/CPlayerInterface.cpp [hunk 3]
```diff
@@ -2510,10 +2510,20 @@ void CPlayerInterface::stacksRebalanced(const StackLocation &src, const StackLoc
 	garrisonsChanged(objects);
 }
 
+void CPlayerInterface::askToAssembleArtifact(const ArtifactLocation &al)
+{
+	auto hero = dynamic_cast<const CGHeroInstance*>(al.relatedObj());
+	if(hero)
+	{
+	  CArtPlace::askToAssemble(hero->getArt(al.slot), al.slot, hero);
```
Comment (alexvins) [inline client/CPlayerInterface.cpp]: One more missing tab and unnecessary {}

File: client/widgets/CArtifactHolder.cpp [hunk 4]
```diff
@@ -225,22 +251,10 @@ void CArtPlace::clickRight(tribool down, bool previousState)
...
-					{
-                        logGlobal->warnStream() << "More than one possibility of assembling... taking only first";
-						break;
-					}
-					return;
-				}
+        if (askToAssemble(ourArt, slotID, ourOwner->curHero))
+        {
+          return;
+        }
```
Comment (alexvins) [inline client/widgets/CArtifactHolder.cpp]: And one more missing tab and unnecessary {}

**Comment by alexvins** [comment]
>to create a blocking InfoWindow with close event handling

There are no blocking dialogs but there is CPlayerInterface::waitWhileDialog(), it might help

**Comment by vmarkovtsev** [comment]
Fixed tabs

**Pull request merged**